File: zig-game-loop/elements.zig
const math = @import("std").math;

const frame = @import("frame.zig");
const ctx = @import("canvas.zig");
const LineCap = ctx.LineCap;
const Path2D = @import("path2d.zig");
const mouse = @import("mouse.zig");

const index = @import("index.zig");
const Cell = index.Cell;
const Connection = index.Connection;
const RotationDirection = index.RotationDirection;
const Color = index.Color;

var hexagon: u32 = 0;
pub const hexagonUnitHeight: f64 = math.sqrt(3.0) / 2.0;

const base03 = 0x002b36;
const base02 = 0x073642;
const base0 = 0x839496;
const red = 0xdc322f;
const blue = 0x268bd2;
const green = 0x859900;
const backgroundColor = base03;
const cellBackgroundColor = base02;
const cellForegroundColor = base0;

pub fn init() void {
    hexagon = createHexagon();
}

fn createHexagon() u32 {
    const path = Path2D.new();
    // workaround for https://github.com/ziglang/zig/issues/12880
    // Path2D.moveTo(path, 0, -1);
    // Path2D.lineTo(path, hexagonUnitHeight, -0.5);
    // Path2D.lineTo(path, hexagonUnitHeight, 0.5);
    // Path2D.lineTo(path, 0, 1);
    // Path2D.lineTo(path, -hexagonUnitHeight, 0.5);
    // Path2D.lineTo(path, -hexagonUnitHeight, -0.5);
    Path2D.m(path, 0, -1);
    Path2D.l(path, hexagonUnitHeight, -0.5);
    Path2D.l(path, hexagonUnitHeight, 0.5);
    Path2D.l(path, 0, 1);
    Path2D.l(path, -hexagonUnitHeight, 0.5);
    Path2D.l(path, -hexagonUnitHeight, -0.5);
    Path2D.closePath(path);
    return path;
}

pub fn background(width: f64, height: f64) void {
    ctx.save();
    defer ctx.restore();
    ctx.fillStyle(backgroundColor);
    ctx.fillRect(0, 0, width, height);
}

fn cellBackground() bool {
    ctx.save();
    defer ctx.restore();
    ctx.scale(0.855, 0.855);
    ctx.fillStyle(cellBackgroundColor);
    ctx.fillPath(hexagon);
    var clicked = false;
    if (frame.time() == mouse.timestamp() and mouse.primary()) {
        const x = mouse.positionX();
        const y = mouse.positionY();
        clicked = ctx.isPointInPath(hexagon, x, y);
    }
    return clicked;
}

fn lerp(t: f64, a: f64, b: f64) f64 {
    return (1 - t) * a + t * b;
}

const rotationAmount: f64 = 60.0;
pub fn cellBackgroundAndEdges(size: i8, cell: Cell) bool {
    _ = size;
    const orientation = cell.orientation;
    const connection = cell.connection;
    var color = cell.color;
    const clicked = cellBackground();

    ctx.save();
    defer ctx.restore();
    const t = lerp(math.min((frame.time() - orientation.startTime) / 250, 1), 0, 1);
    const startAngle =
        @intToFloat(f64, @enumToInt(orientation.value)) + if (orientation.animate == RotationDirection.clockwise) -rotationAmount else rotationAmount;
    const endAngle = @intToFloat(f64, @enumToInt(orientation.value));
    ctx.rotate(lerp(t, startAngle, endAngle) *
        math.pi /
        180);
    if (t < 1) {
        frame.next();
    }
    const rgb: u32 = switch (color) {
        .none => cellForegroundColor,
        .red => red,
        .green => green,
        .blue => blue,
    };
    ctx.fillStyle(rgb);
    ctx.strokeStyle(rgb);
    edges(connection);

    return clicked;
}

fn edges(connection: Connection) void {
    switch (connection) {
        .i => {
            {
                ctx.save();
                defer ctx.restore();
                ctx.lineWidth(0.25);
                ctx.lineCap(LineCap.round);
                ctx.beginPath();
                ctx.moveTo(0.0, 0.0);
                ctx.lineTo(hexagonUnitHeight, 0);
                ctx.stroke();
            }

            ctx.save();
            defer ctx.restore();
            ctx.scale(0.25, 0.25);
            ctx.fillPath(hexagon);
        },
        .v => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .C => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((120.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .l => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .E => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .y => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((120.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .lambda => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .tri => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((120.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((120.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .K => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .phi => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0.0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((120.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((120.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .X => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .hat => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(0, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
        .star => {
            ctx.save();
            defer ctx.restore();
            ctx.lineWidth(0.25);
            ctx.lineCap(LineCap.round);
            ctx.beginPath();
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
            ctx.beginPath();
            ctx.rotate((60.0 * math.pi) / 180.0);
            ctx.moveTo(-hexagonUnitHeight, 0);
            ctx.lineTo(hexagonUnitHeight, 0);
            ctx.stroke();
        },
    }
}

File: ts-game-loop/elements.ts
import { Frame } from "../play/index";
import { Cell, Connection } from "./index";

const base03 = "#002b36";
const base02 = "#073642";
const base0 = "#839496";
const red = "#dc322f";
const blue = "#268bd2";
const green = "#859900";
const backgroundColor = base03;
const cellBackgroundColor = base02;
const cellForegroundColor = base0;

export const hexagonUnitHeight = Math.sqrt(3) / 2;
const hexagon = new Path2D(`
  m ${-hexagonUnitHeight} -0.5
  l  ${hexagonUnitHeight} -0.5
  l  ${hexagonUnitHeight}  0.5
  l                     0    1
  l ${-hexagonUnitHeight}  0.5
  l ${-hexagonUnitHeight} -0.5
  z
`);

export const background = (
	ctx: CanvasRenderingContext2D,
	width: number,
	height: number
) => {
	ctx.save();
	ctx.fillStyle = backgroundColor;
	ctx.fillRect(0, 0, width, height);
	ctx.restore();
};

export const cellBackground = (ctx: CanvasRenderingContext2D): boolean => {
	ctx.save();
	ctx.scale(0.855, 0.855);
	ctx.fillStyle = cellBackgroundColor;
	ctx.fill(hexagon);
	const clicked = false; //ctx.isPointInPath(hexagon, x, y);
	ctx.restore();
	return clicked;
};

const lerp = (t: number, a: number, b: number): number => {
	return (1 - t) * a + t * b;
};

import BezierEasing from "bezier-easing";
const easing = BezierEasing(0.25, 0.1, 0.25, 1);
export const cellBackgroundAndEdges = (
	frame: Frame,
	ctx: CanvasRenderingContext2D,
	{ orientation, connection, color }: Cell
): boolean => {
	const clicked = cellBackground(ctx);

	ctx.save();
	const t = easing(
		Math.min((frame.time() - orientation.startTime) / orientation.duration, 1)
	);
	const startAngle =
		orientation.animate === "clockwise"
			? orientation.value - 60
			: orientation.value + 60;
	const endAngle = orientation.value;
	ctx.rotate((lerp(t, startAngle, endAngle) * Math.PI) / 180);
	if (t < 1) {
		frame.next();
	}
	switch (color) {
		case "none":
			ctx.fillStyle = ctx.strokeStyle = cellForegroundColor;
			break;
		case "red":
			ctx.fillStyle = ctx.strokeStyle = red;
			break;
		case "green":
			ctx.fillStyle = ctx.strokeStyle = green;
			break;
		case "blue":
			ctx.fillStyle = ctx.strokeStyle = blue;
			break;
	}
	edges(ctx, connection);
	ctx.restore();

	return clicked;
};

export const gameOverCell = (
	frame: Frame,
	ctx: CanvasRenderingContext2D,
	{ orientation, connection }: Cell,
	l: number
) => {
	cellBackground(ctx);

	ctx.save();
	const t = easing(
		Math.min((frame.time() - orientation.startTime) / orientation.duration, 1)
	);
	ctx.rotate(
		(lerp(
			t,
			orientation.animate === "clockwise"
				? orientation.value - 60
				: orientation.value + 60,
			orientation.value
		) *
			Math.PI) /
			180
	);
	if (t < 1) {
		frame.next();
	}
	ctx.fillStyle = ctx.strokeStyle = `hsl(51deg 100% ${l}%)`;
	edges(ctx, connection);
	ctx.restore();
};

export const edges = (
	ctx: CanvasRenderingContext2D,
	connection: Connection
) => {
	switch (connection) {
		case 0b100000: // i
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();

			ctx.save();
			ctx.scale(0.25, 0.25);
			ctx.fill(hexagon);
			ctx.restore();
			break;
		case 0b110000: // v
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b101000: // C
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((120 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b100100: // l
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b111000: // E
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b101100: // y
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((120 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b110100: // λ
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b101010: // tri
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((120 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((120 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b111100: // K
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b101110: // Ψ
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((120 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((120 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b110110: // X
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b111110: // hat
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(0, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
		case 0b111111: // star
			ctx.save();
			ctx.lineWidth = 0.25;
			ctx.lineCap = "round";
			ctx.beginPath();
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.beginPath();
			ctx.rotate((60 * Math.PI) / 180);
			ctx.moveTo(-hexagonUnitHeight, 0);
			ctx.lineTo(hexagonUnitHeight, 0);
			ctx.stroke();
			ctx.restore();
			break;
	}
};
